File: JANUARY/10-01-2026/Substrings with K Distinct.cpp
class Solution {
  public:
    int countSubstr(string& s, int k) {
        // // code here
    
        vector<int> freq1(26, 0), freq2(26, 0);
        int cnt1 = 0, cnt2 = 0;
        int m = 0, n = 0;
        int ans = 0;
    
        for (int i = 0; i < s.size(); i++) {
    
            if (freq1[s[i] - 'a']++ == 0)
                cnt1++;
    
            if (freq2[s[i] - 'a']++ == 0)
                cnt2++;
    
            while (cnt1 > k) {
                if (--freq1[s[m] - 'a'] == 0)
                    cnt1--;
                m++;
            }
    
            while (cnt2 > k - 1) {
                if (--freq2[s[n] - 'a'] == 0)
                    cnt2--;
                n++;
            }
    
            ans += (m - n);
        }
    
        return ans;

        
    }
}

File: JANUARY/10-01-2026/Substrings with K Distinct.java
class Solution {
    public int atMost(String s, int k) {
        // code here
        HashMap<Character, Integer> map = new HashMap<>();
        int l = 0, r = 0, cnt = 0;
        int n = s.length();

        while (l < n) {
            map.put(s.charAt(l), map.getOrDefault(s.charAt(l), 0) + 1);

            while (map.size() > k) {
                map.put(s.charAt(r), map.get(s.charAt(r)) - 1);
                if (map.get(s.charAt(r)) == 0)
                    map.remove(s.charAt(r));
                r++;
            }

            cnt += l - r + 1;
            l++;
        }
        return cnt;
    }
    public int countSubstr(String s, int k){
        return atMost(s,k)- atMost(s,k-1);
    }
    
}